First sketch of solution to be discussed.

yakoubix opened the pull request:
We modified the option values at the penultimate node using BlackCalculator class. We added too a boolean variable to the engine in order to implement the new pricing scheme.

lballabio commented:
Hi, the idea is correct.  I'll follow up with a few small corrections.  In the meantime, you can start modifying the `main` so that you can observe the oscillations — that is, so that it calculates the option price for different time steps.

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
         // Rollback to third-last step, and get underlying prices (s2) &
         // option values (p2) at this point
+
+        Real* prices_pen = new Real[timeSteps_ - 1];
+        Real* underlyings_pen = new Real[timeSteps_ - 1];
```
You're allocating memory here, so if you don't delete it afterwards you'll have a memory leak.

However (see below) you don't really need arrays.  You can use local variables when you need them.

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
+            for (int i = 0; i < timeSteps_-2; i++) {
+                underlyings_pen[i] = lattice->underlying(timeSteps_ - 2, i);
+                BlackCalculator bs_option(payoff->optionType(), payoff->strike(), underlyings_pen[i], v*std::sqrt(maturity/timeSteps_), std::exp(-r* maturity/ timeSteps_));
+                Real value_bs = bs_option.value();
+                prices_pen[i] = value_bs;
+                option.values()[i] = prices_pen[i];
```
You can simplify this as
```suggestion
                option.values()[i] = bs_option.value();
```
so you don't need an array.

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
+        Real* underlyings_pen = new Real[timeSteps_ - 1];
+
+        if (oscillations_ == true) {
+            option.rollback(grid[timeSteps_ - 2]);
```
Check the index of the grid here.  It might help if you draw a tree with 4 or 5 steps and you work out the index in that case.

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
+        if (oscillations_ == true) {
+            option.rollback(grid[timeSteps_ - 2]);
+            Array vapenu(option.values());
+            QL_ENSURE(vapenu.size() == timeSteps_ - 1, "Expect time steps minus 1 nodes in grid at penultimate step");
+            for (int i = 0; i < timeSteps_-2; i++) {
```
You're checking that the number of nodes is `timeSteps_ - 1`, but in the next line you loop over `timeSteps_ -2` nodes.  Something is off.

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
+            Array vapenu(option.values());
+            QL_ENSURE(vapenu.size() == timeSteps_ - 1, "Expect time steps minus 1 nodes in grid at penultimate step");
+            for (int i = 0; i < timeSteps_-2; i++) {
+                underlyings_pen[i] = lattice->underlying(timeSteps_ - 2, i);
```
You don't need an array here: you can use a local variable and write `Real u_pen = ...`

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
+            QL_ENSURE(vapenu.size() == timeSteps_ - 1, "Expect time steps minus 1 nodes in grid at penultimate step");
+            for (int i = 0; i < timeSteps_-2; i++) {
+                underlyings_pen[i] = lattice->underlying(timeSteps_ - 2, i);
+                BlackCalculator bs_option(payoff->optionType(), payoff->strike(), underlyings_pen[i], v*std::sqrt(maturity/timeSteps_), std::exp(-r* maturity/ timeSteps_));
```
1) you're using `maturity/timeSteps_` a few times, so it would make sense to calculate it once as `Time dT = maturity/timeSteps_;` and use `dT` in the formulas.
2) the `BlackCalculator` class asks for the forward value of the underlying at maturity, not the underlying.  That would be `underlying * std::exp((r-q)*dT)`.

lballabio reviewed (commented):
A few comments.

yakoubix commented:
Thank you for your recommendations and corrections. I have corrected and pushed a new version of the code implementing them.
Regarding modifying the main in order to observe the oscillations, could you please tell us if we need to show the result output in the form of a summary table ? 
Also, we can see that the put price for time steps greater than 415 for example, takes the value of 4.20359 if it's odd and 4.2036 if it's even. Can we consider this as oscillations ? Knowing that in the table 1 of the projects summary the moves in the option price are more important.

lballabio commented:
A summary table is ok.  For example, in the main you can write a `for` loop over a range of time steps and output the number of steps and the price.

What you're seeing are small oscillations — are you running with `oscillations` equal to `true` or `false`?  Try the other case and compare the size of the oscillations.  If they are too small and they make it hard to see the difference, try using fewer time steps.  You can also try different trees; the current `main` uses Jarrow-Rudd, but there are others.

yakoubix commented:
Hello - I added a summary table output to represent the option value and the time steps and changed the tree type to an AdditiveEQPBinomialTree. The oscillations resolution scheme seems to work pretty well as the option value get stabilized as the time step increase.

lballabio commented:
The implementation looks ok.  There might be some numerical fixes to do — if you plot the values with and without dampened oscillations, the values after the change look a bit low.  I'll try to find some time and check the numbers.  In the meantime, you can double-check it as well.

yakoubix commented:
Hello - I checked the numbers and especially the parameters given to my BlackCalculator but I still cannot find the error that might lead to a lower option value when using the oscillations pricing scheme.  - I updated the code so that it plots the value with and without dampened oscillations. You can find hereafter a plot done using gnuplot that shows the evolution of the option value with the time steps.

[Option_values_oscillations.pdf](https://github.com/lballabio/IMT2021/files/6177866/Option_values_oscillations.pdf)

lballabio commented:
Yes, I checked too, and the parameters look correct.  This method seems to converge to a slightly lower value — we'll take that as a fact of life :)

Good.  You can go ahead with the analysis and the report.

lballabio commented on project4/binomialengine.hpp:
```diff
@@ -128,8 +129,26 @@ namespace QuantLib {
+            for (int i = 0; i < timeSteps_; i++) {
+                Real u_pen = lattice->underlying(timeSteps_ - 1, i);
+                BlackCalculator bs_option(payoff->optionType(), payoff->strike(), u_pen * std::exp((r - q) * dT), v*std::sqrt(dT), std::exp(-r*dT));
+                option.values()[i] = bs_option.value();
```
One last thing: it doesn't change the values significantly, but this should be
```suggestion
                option.values()[i] = std::max(bs_option.value(), (*payoff)(u_pen));
```
because the holder of the option can exercise it and get the payoff if it's worth more then the option value.

yakoubix commented:
Hello - I updated with the final version of the code. I omitted the fact that the option type is American and that it can be exercised at every node of the tree. We will start writing the analysis and the report (Readme) accordingly.

yakoubix commented:
Hello - We pushed our final version of the project 4 including the final report.

lballabio commented:
Thanks.  If you still have time, you might check what happens with different kinds of trees.  But otherwise, the report can stand as it is.

yakoubix commented:
We pushed a second final version of the report including the results analysis of some different tree types (CRR & Tian). Thank you.

lballabio commented:
Thanks!
